### PR TITLE
tests: Introduce test writing guidelines

### PR DESCRIPTION
**Pull request by robertbaldyga**
Signed-off-by: Robert Baldyga <robert.baldyga@intel.com>

### CODE REVIEW

File: test/functional/GUIDELINES.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+This document contains guidlines and BKM for writing tests. Following these
+will significantly reduce numbers of comments you will receive on the code
+review and will speed up process of merging your patch.
+
+# Tests structure
+
+- All the tests reside in `tests/` directory, which contains subdirectories
+corresponding to individual test groups.
+- Test gruup directories contain tests that belong to particular group.
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: group

File: test/functional/GUIDELINES.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+Each test should belong to exactly one test group.
+- Each test shoud be placed in separate file (although test may comprise many
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: should

File: test/functional/GUIDELINES.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+test cases).
+
+# Test file
+
+- A test file corresponds to individual test.
+- Test file name should be preceded with `test_` prefix, which allows pytest
+to collect it using standard test discovery method.
+- Test file should contain one or more test cases.
+- Each test case is an function with the name preceded with `test_` prefix.
```
Comment (katlapinka) [inline test/functional/GUIDELINES.md]: is a function

File: test/functional/GUIDELINES.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+Please refer to Test Framework documentation (coming soon).
+
+## Testing code
+
+- Typical test case comprises three basic elements: preparation, testing
+logic, and cleanup.
+- Test Framework allows to define test steps, step groups and iterations.
+- Every single line of test code must belong to some test step.
+- Test steps may be embedded into test step grups and iteration.
```
Comment (katlapinka) [inline test/functional/GUIDELINES.md]: groups ans interations

File: test/functional/GUIDELINES.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+- Test grups may be embedded into another test groups and iterations.
```
Comment (katlapinka) [inline test/functional/GUIDELINES.md]: groups

Comment (Deixx) [inline test/functional/GUIDELINES.md]: other test groups
(another is singular)

File: test/functional/GUIDELINES.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+- Iterations may be embedded into test groups ant another iterations.
+- Neither test steps nor test grups nor iterations may be embedded into
```
Comment (katlapinka) [inline test/functional/GUIDELINES.md]: group

File: test/functional/GUIDELINES.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+## Testing code
+
+- Typical test case comprises three basic elements: preparation, testing
+logic, and cleanup.
+- Test Framework allows to define test steps, step groups and iterations.
+- Every single line of test code must belong to some test step.
+- Test steps may be embedded into test step grups and iteration.
+- Test grups may be embedded into another test groups and iterations.
+- Iterations may be embedded into test groups ant another iterations.
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: and other

File: test/functional/GUIDELINES.md (hunk 3 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Fixed.

File: test/functional/GUIDELINES.md (hunk 4 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Fixed.

File: test/functional/GUIDELINES.md (hunk 5 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Fixed.

File: test/functional/GUIDELINES.md (hunk 6 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Fixed.

File: test/functional/GUIDELINES.md (hunk 7 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Fixed.

File: test/functional/GUIDELINES.md (hunk 5 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Fixed.

File: test/functional/GUIDELINES.md (hunk 2 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Fixed.

File: test/functional/GUIDELINES.md (hunk 1 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Fixed.

File: test/functional/GUIDELINES.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+def test_example():
+    """
+    title: Brief test title
+    description: |
+      More detailed test description.
+      This is second line of test description.
+    pass_criteria:
+      - First pass criterium
+      - Second pass criterium
+      - Third pass criterium
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: The first, etc.

File: test/functional/GUIDELINES.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    """
+```
+
+## Pytest marks
+
+Test cases can use both standard pytest marks like `@pytest.mark.parametrize`
+as well as marks introduced by test framework like `@pytest.mark.require_disk`.
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: the test framework

File: test/functional/GUIDELINES.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+Please refer to Test Framework documentation (coming soon).
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: the Test Framework

File: test/functional/GUIDELINES.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+## Testing code
+
+- Typical test case comprises three basic elements: preparation, testing
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: A typical

File: test/functional/GUIDELINES.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+logic, and cleanup.
+- Test Framework allows to define test steps, step groups and iterations.
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: The test framework

File: test/functional/GUIDELINES.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+- Every single line of test code must belong to some test step.
+- Test steps may be embedded into test step groups and iterations.
+- Test groups may be embedded into other test groups and iterations.
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: Test step groups ... into other test step groups (test groups have different meaning)

File: test/functional/GUIDELINES.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+- Iterations may be embedded into test groups and other iterations.
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: into test step groups

File: test/functional/GUIDELINES.md [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+- Neither test steps nor test groups nor iterations may be embedded into
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: nor test step groups

File: test/functional/GUIDELINES.md [hunk 16]
```diff
@@ -0,0 +1,80 @@
+# Preface
+
+This document contains guidlines and BKM for writing tests. Following these
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+will significantly reduce numbers of comments you will receive on the code
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+review and will speed up process of merging your patch.
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+# Tests structure
+
+- All the tests reside in `tests/` directory, which contains subdirectories
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+corresponding to individual test groups.
+- Test group directories contain tests that belong to particular group.
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+Each test should belong to exactly one test group.
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: No, it was not.

File: test/functional/GUIDELINES.md [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+- Each test should be placed in separate file (although test may comprise many
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+test cases).
+
+# Test file
+
+- A test file corresponds to individual test.
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+- Test file name should be preceded with `test_` prefix, which allows pytest
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+to collect it using standard test discovery method.
+- Test file should contain one or more test cases.
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+- Each test case is a function with the name preceded with `test_` prefix.
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+- Test file can also contain another functions, classes etc. used by the test
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+# Test case
+
+A test case is a single function. There are four fundamental elements of each
+test function:
+1. Function name
+2. Docstring
+3. Pytest marks
+4. Testing code
+
+Following sections describe guidlines for each of these elements.
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 29]
```diff
@@ -0,0 +1,80 @@
...
+
+## Function name
+
+- The name of the test case function should begin with `test_` prefix.
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 30]
```diff
@@ -0,0 +1,80 @@
...
+## Docstring
+
+Each test case should contain yaml docstring in the following format:
+```python
+def test_example():
+    """
+    title: Brief test title
+    description: |
+      More detailed test description.
+      This is second line of test description.
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 31]
```diff
@@ -0,0 +1,80 @@
...
+
+## Docstring
+
+Each test case should contain yaml docstring in the following format:
+```python
+def test_example():
+    """
+    title: Brief test title
+    description: |
+      More detailed test description.
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 32]
```diff
@@ -0,0 +1,80 @@
...
+- The name should be consise and meaningful in terms of what the test case
+actually does.
+
+## Docstring
+
+Each test case should contain yaml docstring in the following format:
+```python
+def test_example():
+    """
+    title: Brief test title
```
[the remark this answers is not in the document]
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md (hunk 8 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md (hunk 9 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md (hunk 11 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md (hunk 12 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md (hunk 14 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md (hunk 15 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md (hunk 10 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md (hunk 13 above)
Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

File: test/functional/GUIDELINES.md [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+- Step groups may be embedded into other step groups and iterations.
+- Iterations may be embedded into step groups and other iterations.
+- Neither test steps nor step groups nor iterations may be embedded into
+  test steps. In other words, test steps must be the deepest elements in
+  the test structure and they may contain only the testing code.
+
+### Test implementation
+
+- Avoid using `TestRun.executor` directly. It's very likely that the Test
+  Framework or CAS API already has functionality you need. If not, then you
```
Comment (Deixx) [inline test/functional/GUIDELINES.md]: the functionality

Comment (robertbaldyga) [reply test/functional/GUIDELINES.md]: Done.

**Pull request merged**